aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -0,0 +1,281 @@
+        };
+        // re-insert items from the splitted bucket
+        let i = 0;
+        let len = Vector::length(&old_bucket);
+        while (i < len) {
+            let entry = Vector::pop_back(&mut old_bucket);
+            let index = bucket_index(map, entry.hash);
+            Vector::push_back(Table::borrow_mut(&mut map.buckets, &index), entry);
+            i = i + 1;
+        };
+        Vector::destroy_empty(old_bucket);
+    }
+
+    /// Return the expected bucket index to find the hash.
+    fun bucket_index<K, V>(map:& HashMap<K, V>, hash: u64): u64 {
+        let index = hash % (1 << (map.level + 1));
+        if (index < map.num_buckets) {
+            // in existing bucket
+            index
+        } else {
+            // in unsplitted bucket
+            index % (1 << map.level)
+        }
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    /// The requirement of &mut HashMap is to bypass the borrow checker issue described in https://github.com/move-language/move/issues/95
+    public fun borrow<K, V>(map: &mut HashMap<K, V>, key: &K): &V {
+        let index = bucket_index(map, sip_hash(key));
+        let bucket = Table::borrow_mut(&mut map.buckets, &index);
+        let i = 0;
+        let len = Vector::length(bucket);
+        while (i < len) {
+            let entry = Vector::borrow(bucket, i);
+            if (&entry.key == key) {
+                return &entry.value
+            };
+            i = i + 1;
+        };
+        abort Errors::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K, V>(map: &mut HashMap<K, V>, key: &K): &mut V {
+        let index = bucket_index(map, sip_hash(key));
+        let bucket = Table::borrow_mut(&mut map.buckets, &index);
+        let i = 0;
+        let len = Vector::length(bucket);
+        while (i < len) {
+            let entry = Vector::borrow_mut(bucket, i);
+            if (&entry.key == key) {
+                return &mut entry.value
+            };
+            i = i + 1;
+        };
+        abort Errors::invalid_argument(ENOT_FOUND)
+    }
+
+    /// Returns true iff `table` contains an entry for `key`.
+    public fun contains<K, V>(map: &HashMap<K, V>, key: &K): bool {
+        let index = bucket_index(map, sip_hash(key));
+        let bucket = Table::borrow(&map.buckets, &index);
+        let i = 0;
+        let len = Vector::length(bucket);
+        while (i < len) {
+            let entry = Vector::borrow(bucket, i);
+            if (&entry.key == key) {
+                return true
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    /// Remove from `table` and return the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun remove<K: drop, V>(map: &mut HashMap<K,V>, key: &K): V {
```

Comment:
it's easy to add shrink_one_bucket if needed

Reply:
I would prefer we add `shrink_one_bucket` if that's the case to avoid having tables with too few items but with a large number of buckets - thoughts?